xgcm/autogenerate.py

```diff
@@ -1,254 +1,240 @@
-                                                         boundary=boundary,
-                                                         fill_value=fill_value)
+from .grid import Axis
+
+
+def generate_axis(ds,
+                  axis,
+                  name,
+                  axis_dim,
+                  pos_from='center',
+                  pos_to='left',
+                  wrap=None,
+                  pad=None,
+                  raw_switch=True):
+    """
+    Parameters
+    ----------
+    ds : xarray.Dataset
+        Dataset with gridinformation used to construct c-grid
+    axis_dict : dict
+        Dict with information on the dimension in ds corrsponding to the xgcm
+        axis. E.g. {'X':'lon'}
+    pos_from : str
+        Position of the gridpoints given in 'ds'. Can be
+        {'center','left','right'}
+    pos_to : str
+        Position of the gridpoints to be generated. Can be
+        {'center','left','right'}
+    wrap : None or float
+        If specified, marks the value of discontinuity across boundary, e.g.
+        360 for global longitude values and 180 for global latitudes.
+    pad : {None, float, 'auto'}
+        If specified, determines the padding to be applied across boundary.
+        If float is specified, that value is used as padding. Auto attempts to
+        pad linearly extrapolated values. Can be useful for e.g. depth
+        coordinates (to reconstruct 0 depth). Can lead to unexpected values
+        when coordinate is multidimensional.
+    raw_switch : bool
+        Determines if the attributes are created from scratch. Should be
+        enabled for dimensions and deactivated for multidimensional
+        coordinates. These can only be calculated after the dims are created.
+    """
+    new_name = name+'_inferred'
+
+    # Determine the relative position to interpolate to based on current and
+    # desired position
+    relative_pos_to = position_to_relative(pos_from, pos_to)
+
+    if (wrap is not None) and (pad is not None):
+        raise RuntimeError('Coordinate cannot be wrapped and padded at the\
+                            same time')
+    elif (wrap is None) and (pad is not None):
+        if pad == 'auto':
+            left_pad, right_pad = auto_pad(ds[name], axis_dim)
+            if relative_pos_to == 'right':
+                fill_value = right_pad
+            elif relative_pos_to == 'left':
+                fill_value = left_pad
+        else:
+            fill_value = pad
+        periodic = False
+        boundary = 'fill'
+    elif (wrap is not None) and (pad is None):
+        periodic = True
+        fill_value = 0.0
+        boundary = None
+    else:
+        raise RuntimeError('Either "wrap" or "pad" have to be specified')
+
+    if raw_switch:
+        # Input coordinate has to be declared as center,
+        # or xgcm.Axis throws error. Will be rewrapped below.
+        ds[name] = fill_attrs(ds[name], 'center', axis)
```

Comment:
I would probably copy `ds` before adding new variables to it

Reply:
Good call, that caused some issues in the tests earlier. Added `ds=ds.copy` before